### PR TITLE
Adding Decoders for F5 Big IP APM and AFM

### LINKED ISSUE #15848: Create F5 decoder
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
|5.0.0 | Engine | Manager | Packages/Sources | OS version |

## Description
It is necessary to extend the coverage of our ruleset, so a F5 decoder will be written for this purpose.

## DoD

- [x] Create decoder files.
- [x] Perform Validation and add to a testing environment.
- [x] Define testing policy. Manual testing.
- [x] Documentation.

### PR DESCRIPTION
**Pull request by JorgeMarinoDev**
|Related Issue|
|---|
| #15848|

## Description

With the aim of expanding the coverage of the decoders, this PR adds a decoder for F5 BigIp APM and AFM.

## Continue work from #16054

After initial reviews there were many changes made and a pair of files were unified.  This makes the review process hard to follow.
All comments in the initial reviews were took in consideration.

## Configuration options

**Validate the new decoders**:

- main catalog validate decoder/f5-apm/0 < engine/ruleset/decoders/f5/f5-apm.yml
- main catalog validate decoder/f5-afm/0 < engine/ruleset/decoders/f5/f5-afm.yml

**Create decoders**:

- main catalog create decoder < engine/ruleset/decoders/f5/f5-apm.yml
- main catalog create decoder < engine/ruleset/decoders/f5/f5-afm.yml

**Update environment**:

- main catalog update environment/wazuh/0 < engine/ruleset/environments/wazuh-environment.yml

**Test new decoder**
- ./build/main test -q -1

## Tests

Event:
iusm modtempo olab6078.home olaboris tur itv [F5@odoco acl_policy_name=ria acl_policy_type=min acl_rule_name=ite action=Closed hostname=tatemac3541.api.corp bigip_mgmt_ip=10.228.193.207 context_name=liqua context_type=ciade date_time=Jan 29 2016 06:09:59 dest_ip=10.125.114.51 dst_geo=umq dest_port=2288 device_product=pexe device_vendor=nes device_version=1.2262 drop_reason=reveri errdefs_msgno=boNemoe errdefs_msg_name=equepor flow_id=eni ip_protocol=ipv6 severity=low partition_name=ehend route_domain=ritquiin sa_translation_pool=umqui sa_translation_type=reeufugi source_ip=10.208.121.85 src_geo=sperna source_port=884 source_user=billoi translated_dest_ip=10.165.201.71 translated_dest_port=6153 translated_ip_protocol=tatemU translated_route_domain=deF translated_source_ip=10.11.196.142 translated_source_port=5222 translated_vlan=iatnu vlan=3810

Output:
```
{
    "wazuh": {
        "queue": 49,
        "origin": "/dev/stdin"
    },
    "event": {
        "original": "iusm modtempo olab6078.home olaboris tur itv [F5@odoco acl_policy_name=ria acl_policy_type=min acl_rule_name=ite action=Closed hostname=tatemac3541.api.corp bigip_mgmt_ip=10.228.193.207 context_name=liqua context_type=ciade date_time=Jan 29 2016 06:09:59 dest_ip=10.125.114.51 dst_geo=umq dest_port=2288 device_product=pexe device_vendor=nes device_version=1.2262 drop_reason=reveri errdefs_msgno=boNemoe errdefs_msg_name=equepor flow_id=eni ip_protocol=ipv6 severity=low partition_name=ehend route_domain=ritquiin sa_translation_pool=umqui sa_translation_type=reeufugi source_ip=10.208.121.85 src_geo=sperna source_port=884 source_user=billoi translated_dest_ip=10.165.201.71 translated_dest_port=6153 translated_ip_protocol=tatemU translated_route_domain=deF translated_source_ip=10.11.196.142 translated_source_port=5222 translated_vlan=iatnu vlan=3810",
        "action": "Closed",
        "code": "boNemoe",
        "dataset": "f5.bigipafm",
        "module": "f5"
    },
    "@timestamp": "2016-01-29T06:09:59.000Z",
    "destination": {
        "geo": {
            "country_name": "umq"
        },
        "nat": {
            "ip": "10.165.201.71",
            "port": 6153
        },
        "port": 2288
    },
    "fileset": {
        "name": "bigipafm"
    },
    "host": {
        "ip": "10.228.193.207",
        "name": "tatemac3541.api.corp"
    },
    "input": {
        "type": "log"
    },
    "log": {
        "level": "low"
    },
    "network": {
        "protocol": "ipv6"
    },
    "observer": {
        "product": "pexe",
        "type": "Firewall",
        "vendor": "F5",
        "version": 1.2262
    },
    "related": {
        "hosts": [
            "tatemac3541.api.corp"
        ],
        "ip": [
            "10.228.193.207",
            "10.208.121.85",
            "10.165.201.71",
            "10.11.196.142"
        ],
        "user": [
            "billoi"
        ]
    },
    "rsa": {
        "internal": {
            "messageid": "BIGIP_AFM"
        },
        "investigations": {
            "ec_activity": "Disable",
            "ec_subject": "NetworkComm",
            "ec_theme": "Communication"
        },
        "misc": {
            "action": [
                "Closed"
            ],
            "context": "liqua",
            "event_type": "equepor",
            "obj_name": "odoco",
            "policy_name": "ria",
            "rule_name": "ite",
            "severity": "low",
            "version": 1.2262
        },
        "network": {
            "alias_host": [
                "tatemac3541.api.corp"
            ],
            "vlan": 3810
        },
        "time": {
            "event_time_str": "2016-01-29T06:09:59.000Z"
        }
    },
    "rule": {
        "name": "ite"
    },
    "service": {
        "type": "f5"
    },
    "source": {
        "geo": {
            "country_name": "sperna"
        },
        "ip": "10.208.121.85",
        "nat": {
            "ip": "10.11.196.142",
            "port": 5222
        },
        "port": 884
    },
    "tags": [
        "f5.bigipafm",
        "forwarded"
    ],
    "user": {
        "name": "billoi"
    }
}



```

### CODE REVIEW

File: src/engine/ruleset/decoders/f5/f5-afm.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    - ~tmp.kvMap : +parse_kv/$~tmp.kv/=/ /'/'
+    - destination.geo.country_name : $~tmp.kvMap.dst_geo
+    - destination.nat.ip : $~tmp.kvMap.translated_dest_ip
+    - destination.nat.port : $~tmp.kvMap.translated_dest_port
+    - destination.port : $~tmp.kvMap.dest_port
+    - event.action : $~tmp.kvMap.action
+    - event.code : $~tmp.kvMap.errdefs_msgno
+    - event.dataset : f5.bigipafm
+    - event.module : f5
+    - event.provider: +ef_delete
```
Comment (soynof) [inline src/engine/ruleset/decoders/f5/f5-afm.yml:37]: why?

File: src/engine/ruleset/decoders/f5/f5-afm.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    - source.geo.country_name : $~tmp.kvMap.src_geo
+    - source.ip : $~tmp.kvMap.source_ip
+    - source.nat.ip : $~tmp.kvMap.translated_source_ip
+    - source.nat.port : $~tmp.kvMap.translated_source_port
+    - source.port : $~tmp.kvMap.source_port
+    - tags : +a_append/$event.dataset
+    - tags : +a_append/forwarded
+    - user.name : $~tmp.kvMap.source_user
+    - ~tmp: +ef_delete
+    - ~payload : +ef_delete
```
Comment (soynof) [inline src/engine/ruleset/decoders/f5/f5-afm.yml:80]: is necesary?

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+parse:
+  logpar:
+    # April 2017/04/02 01:27:07 caboNem medium laudan[7589]: 01490107: :oconse: mag: AD module: authentication with 'tob' failed: Client 'dolores2519.mail.host' not found in Kerberos database, principal name:deF itempo
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~/ignore/ ><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~/ignore/:><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~>:<~/ignore/ ><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>:<~/ignore/:> <event.provider>:<~/ignore/:><~content>
+
+    # Rule
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:31]: What do you mean with rule?

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+definitions:
+  def_session : :<~>:<~/ignore/ ><~tmp.rsa.misc.log_session_id>
+
+parse:
+  logpar:
+    # April 2017/04/02 01:27:07 caboNem medium laudan[7589]: 01490107: :oconse: mag: AD module: authentication with 'tob' failed: Client 'dolores2519.mail.host' not found in Kerberos database, principal name:deF itempo
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~/ignore/ ><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~/ignore/:><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>[<process.pid>]<~/ignore/:> <event.provider>:<~>:<~/ignore/ ><~content>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~>:<~/ignore/:> <event.provider>:<~/ignore/:><~content>
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:29]: Can we have at least one example per case, please?

File: src/engine/ruleset/decoders/f5/f5-afm.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  module: F5
+  title: F5 AFM Decoder
+  description: Decodes F5 BIG-IP Advanced Firewall Manager logs
+  compatibility: >
+    TBD
+  author:
+    name: Wazuh, Inc.
+    date: 2023/01/10
+  references:
+    - TBD
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-afm.yml:14]: Please replace the "TBD"

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+  module: F5
+  title: F5 APM Decoder
+  description: Decodes F5 BIG-IP Access Policy Manager logs
+  compatibility: >
+    TBD
+  author:
+    name: Wazuh, Inc.
+    date: 2023/02/04
+  references:
+    - TBD
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:14]: Please replace the "TBD"

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 7]
```diff
@@ -23,36 +23,34 @@ definitions:
...
     # November 2019/11/01 10:16:48 erc low tasnu: [syslog-ng]
-    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~tmp.rsa_misc_client>:<~/ignore/:> [<event.provider>]
-    # June 2018/06/04 20:44:15 lorumw high tdolo[3872]: syslog-ng:
-    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~tmp.rsa_misc_client>[<process.pid>]:<~/ignore/ ><event.provider>:<~/ignore/ >
-    # April 2018/04/08 16:33:58 liq low mvolupta: syslog-ng:
-    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~tmp.rsa_misc_client>:<~/ignore/:> <event.provider>:<~/ignore/ >
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <~tmp.rsa_misc_client>:<~/ignore/:> [<event.code>]
     # March 2017/03/18 18:24:33 ptasnula high syslog-ng[2638]: ill
-    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <event.provider>[<process.pid>]:<~/ignore/ ><~tmp.rsa_misc_client>
+    - event.original: <~> <@timestamp/%Y\/%m\/%d %T> <~> <log.level> <event.code>[<process.pid>]:<~/ignore/ ><~tmp.rsa_misc_client>
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:47]: What is it that is being ignored before the Timestamp?

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/ruleset/decoders/f5/f5-apm.yml (hunk 7 above)
Comment (JorgeMarinoDev) [reply src/engine/ruleset/decoders/f5/f5-apm.yml:47]: It's the month string representation

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    date: 2023/02/04
+  references:
+    - https://techdocs.f5.com/kb/en-us/products/big-ip_apm/manuals/product/apm-network-access-12-1-0/10.html
+
+sources:
+  - decoder/queue-syslog/0
+  - decoder/queue-localfile/0
+
+definitions:
+  def_session : :<~>:<~/ignore/ ><~tmp.rsa.misc.log_session_id>
```
Comment (LucioDonda) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:21]: In the other parsers where we use definitions, the names were written in uppercase, It isn't mandatory and it wasn't even defined as good practice, but I guess that it could help users while reading.

**Review by LucioDonda — COMMENTED** [review]
LGTM, but waiting for those corrections, specially the ones related to order of logpar expressions

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  # General fields mapping
+  - map:
+    - event.dataset : f5.bigipapm
+    - event.module : f5
+    - fileset.name : bigipapm
+    - input.type : log
+    - observer.product : Big-IP
+    - observer.type : Access
+    - observer.vendor : F5
+    - rsa.internal.messageid : $event.code
```
[the remark this answers is not in the document]
Comment (JorgeMarinoDev) [reply src/engine/ruleset/decoders/f5/f5-apm.yml:401]: We are not sure. This is redundant, even though found in the reference pipelines. Deleting.

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    - observer.type: Access
+    - observer.vendor: F5
+    - rsa.internal.messageid: $event.code
+    - rsa.misc.log_session_id: $~tmp.rsa.misc.log_session_id
+    - rsa.misc.severity: $log.level
+    - rsa.time.event_time: $@timestamp
+    - service.type: f5
+    - tags: [f5.bigipapm, forwarded]
+    - ~content: +ef_delete
+    - ~tmp: +ef_delete
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:375]: Please, fix the indentation

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/ruleset/decoders/f5/f5-apm.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (JorgeMarinoDev) [reply src/engine/ruleset/decoders/f5/f5-apm.yml]: According to the reference pipelines. Yes. Redundant, yes. We'll keep an eye on this.

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+    map:
+      - url: +ef_rename/$~url
+      - rsa.misc.result: $~tmp.rsa_misc_result
+
+  # 01490106
+  # oremagn[2121]: 01490106: :itseddo: uptatev: AD module: authentication with 'oditem' failed in allow: Preauthentication failed, principal name: inimaven. failure olor
+  - check: event.code==01490106
+    logpar:
+      - ~content: $DEF_SESSION:<~/ignore/:> <~>:<~/ignore/:> <~>'<~tmp.user>' <~>:<~/ignore/:> <~>, <~>. <~tmp.rsa_misc_result> <~>
+    map:
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:63]: The way the examples are displayed might be confusing. Which part of the log is the `~content`?

Comment (nmkoremblum) [reply src/engine/ruleset/decoders/f5/f5-apm.yml:63]: Same observation for all the cases.

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 12]
```diff
@@ -135,7 +148,8 @@ normalize:
       - rsa.misc.action: +a_append/$~tmp.rsa.misc.action
 
   # 01490007
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:150]: I would also remove these comments with the code, as it is explicitly checked on the `check` condition

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+    map:
+      - rsa.internal.event_desc: $~tmp.rsa.internal.event_desc
+      - rsa.web.fqdn: $~tmp.rsa.web.fqdn
+      - related.hosts: +a_append/$rsa.web.fqdn
+
+  # original : May 2016/05/22 14:30:33 metcon low emeumfug[6823]: 01490505: :emporinc: untutlab: tem
+  # content  : :emporinc: untutlab: tem
+  - check: event.code==01490505
+    logpar:
+      - ~content: $DEF_SESSION:<~/ignore/ ><~>
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:245]: Can't we just ignore everything after the `:`?

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+    logpar:
+      - ~content: $DEF_SESSION:<~/ignore/ ><~>, SERVER :<~/ignore/  ><~tmp.rsa.db.index>
+    map:
+      - rsa.db.index: $~tmp.rsa.db.index
+
+  # original : October 2018/10/11 12:07:23 aborio low setquas: 014d0002: :nbyCi: runtmoll: SSOv2 Logon failed, config busBon form norumetM
+  # content  : :nbyCi: runtmoll: SSOv2 Logon failed, config busBon form norumetM
+  - check: event.code==014d0002
+    logpar:
+      - ~content: <~/ignore/ >$DEF_SESSION:<~/ignore/ ><~>
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:327]: Can't we just ignore everything after the `:`?

File: src/engine/ruleset/decoders/f5/f5-apm.yml [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+    logpar:
+      - ~content: :<~>:<~/ignore/ ><~tmp.rsa.db.index>
+    map:
+      - rsa.db.index: $~tmp.rsa.db.index
+
+  # original : September 2019/09/19 13:09:05 psumqui high ncu: 01490079: :quaturve: ciad: Access policy 'diconseq' configuration has changed.Access profile 'utod' configuration changes need to be applied for the new configuration
+  # content  :  :quaturve: ciad: Access policy 'diconseq' configuration has changed.Access profile 'utod' configuration changes need to be applied for the new configuration
+  - check: event.code==01490079
+    logpar:
+      - ~content: <~/ignore/ >$DEF_SESSION:<~>
```
Comment (nmkoremblum) [inline src/engine/ruleset/decoders/f5/f5-apm.yml:343]: (Just as here)

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

**Review by nmkoremblum — APPROVED** [review]
LGTM

**Review by juliancnn — CHANGES_REQUESTED** [review]
These decoders use fields that are not in the schema.

**Comment by JorgeMarinoDev** [comment]
> These decoders use fields that are not in the schema.

F5 BigIp APM and AFM adheres to the RSA community Standards, therefore existing collectors expect this information to be present in these fields. This means, we should ingest it under these conditions. 

Unless we propose a homomorphic mapping that relies on the adopted standard's semantics there is the risk of mislead the collected information.

We are aware of this condition @juliancnn . Should we add these RSA fields to the current schema?

**Review by LucioDonda — APPROVED** [review]
LGTM after than minor change

**Review by cborla — APPROVED** [review]
LGTM

File: src/engine/ruleset/decoders/f5/f5-afm.yml [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    - https://support.f5.com/csp/article/K05327372
+
+sources:
+  - decoder/queue-syslog/0
+  - decoder/queue-localfile/0
+
+parse:
+  logpar:
+    # iusm modtempo olab6078.home olaboris tur itv [F5@odoco acl_policy_name=ria acl_policy_type=min acl_rule_name=ite action=Closed hostname=tatemac3541.api.corp bigip_mgmt_ip=10.228.193.207 context_name=liqua context_type=ciade date_time=Jan 29 2016 06:09:59 dest_ip=10.125.114.51 dst_geo=umq dest_port=2288 device_product=pexe device_vendor=nes device_version=1.2262 drop_reason=reveri errdefs_msgno=boNemoe errdefs_msg_name=equepor flow_id=eni ip_protocol=ipv6 severity=low partition_name=ehend route_domain=ritquiin sa_translation_pool=umqui sa_translation_type=reeufugi source_ip=10.208.121.85 src_geo=sperna source_port=884 source_user=billoi translated_dest_ip=10.165.201.71 translated_dest_port=6153 translated_ip_protocol=tatemU translated_route_domain=deF translated_source_ip=10.11.196.142 translated_source_port=5222 translated_vlan=iatnu vlan=3810
+    - event.original: <~>F5@<~tmp.rsa_obj_name> <~tmp.kv1> date_time=<@timestamp/%b %d %Y %H:%M:%S> <~tmp.kv2>
```
Comment (juliancnn) [inline src/engine/ruleset/decoders/f5/f5-afm.yml:23]: Could you change the @timestamp for event.created?

**Review by juliancnn — CHANGES_REQUESTED** [review]

**Review by NahuFigueroa97 — APPROVED** [review]

File: src/engine/ruleset/decoders/f5/f5-afm.yml (hunk 16 above)
Comment (juliancnn) [reply src/engine/ruleset/decoders/f5/f5-afm.yml:23]: My bad, you can ignore this. LGTM!

**Review by soynof — APPROVED** [review]
It's OK!

**Review by JcabreraC — APPROVED** [review]
LGTM !
